### PR TITLE
[BEAM-3356] Add Go SDK int and varint custom coders

### PR DESCRIPTION
**Pull request by herohde**
* There coders are more reasonable defaults for integers.

### CODE REVIEW

**Comment by herohde** [comment]
R: @wcn3

File: sdks/go/pkg/beam/coder.go [hunk 1]
```diff
@@ -85,18 +85,31 @@ func inferCoder(t FullType) (*coder.Coder, error) {
...
+		case reflectx.Uint, reflectx.Uint8, reflectx.Uint16, reflectx.Uint32, reflectx.Uint64:
+			c, err := coder.NewVarUintZ(t.Type())
+			if err != nil {
+				return nil, err
+			}
+			return &coder.Coder{Kind: coder.Custom, T: t, Custom: c}, nil
 		case reflectx.String, reflectx.ByteSlice:
+			// The type conversions here are very conservative. We handle bytes/strings
+			// equivalently because they are essentially equivalent in the language.
+			// Notably, we do not (currently) support equivalences in numeric data types
```
Comment (wcn3) [inline sdks/go/pkg/beam/coder.go]: This part of the comment would go better with the int conversion code.

Do we want to add support for encoding floats? Some applications associate a weight or score with a value that is float-y, so supporting that easily would be helpful.

Comment (herohde) [inline sdks/go/pkg/beam/coder.go]: > **wcn3** wrote:
> This part of the comment would go better with the int conversion code.
> 
> Do we want to add support for encoding floats? Some applications associate a weight or score with a value that is float-y, so supporting that easily would be helpful.


Rephrased the whole comment. It seemed a bit off, because there is no a priori motivation to pick a coder that needs runtime conversions.

I'm not opposed to be adding floats in a separate PR. Is there a standard binary encoding we can use?

Comment (herohde) [reply sdks/go/pkg/beam/coder.go]: Rephrased the whole comment. It seemed a bit off, because there is no a priori motivation to pick a coder that needs runtime conversions.

I'm not opposed to be adding floats in a separate PR. Is there a standard binary encoding we can use?

**Comment by herohde** [comment]
R: @kennknowles (merge)

**Comment by robertwb** [comment]
LGTM. We need to be careful to not call the zig-zag encoding "VarInt" to avoid confusion with the VarInt used elsewhere.
